### PR TITLE
FS-1215 Add default download option for unsupported firmware

### PR DESCRIPTION
**Pull request by coffeefreak101**
Firmware-syncer only supports downloading from specific vendors. Firmware from other vendors are ignored.

This change adds an option to provide a URL (via env var SYNCER_DEFAULT_DOWNLOAD_URL) as a default location to download firmware from unsupported vendors.

The firmware should be available at the location of the URL. For example, if the `SYNCER_DEFAULT_DOWNLOAD_URL` is `http://example.com/firmware` and the name of the firmware file is `some_firmware.bin` then the firmware should be accessible at `http://example.com/firmware/some_firmware.bin`.

If the env var `SYNCER_DEFAULT_DOWNLOAD_URL` is not set, then firmware-syncer goes back to the previous behavior of logging the vendor as unsupported.

### CODE REVIEW

File: internal/vendors/downloader.go [hunk 1]
```diff
@@ -367,3 +370,64 @@ func (s *S3Downloader) Download(ctx context.Context, downloadDir string, firmwar
...
+	}
+}
+
+// Download will download the given firmware into the given downloadDir,
+// and return the full path to the downloaded file.
+// The file will be downloaded from the sourceURL provided to the SourceOverrideDownloader
+// instead of the firmware's UpstreamURL.
+func (d *SourceOverrideDownloader) Download(ctx context.Context, downloadDir string, firmware *serverservice.ComponentFirmwareVersion) (string, error) {
+	firmwareURL := d.baseURL + firmware.Filename
+	filePath := downloadDir + firmware.Filename
```
Comment (joelrebel) [inline internal/vendors/downloader.go:401]: use `filepath.Join()` instead?

File: internal/vendors/downloader.go [hunk 2]
```diff
@@ -367,3 +370,64 @@ func (s *S3Downloader) Download(ctx context.Context, downloadDir string, firmwar
...
+		sourceURL,
+	}
+}
+
+// Download will download the given firmware into the given downloadDir,
+// and return the full path to the downloaded file.
+// The file will be downloaded from the sourceURL provided to the SourceOverrideDownloader
+// instead of the firmware's UpstreamURL.
+func (d *SourceOverrideDownloader) Download(ctx context.Context, downloadDir string, firmware *serverservice.ComponentFirmwareVersion) (string, error) {
+	firmwareURL := d.baseURL + firmware.Filename
```
Comment (joelrebel) [inline internal/vendors/downloader.go:400]: use `url.JoinPath()` instead?

File: internal/vendors/downloader.go [hunk 3]
```diff
@@ -367,3 +370,64 @@ func (s *S3Downloader) Download(ctx context.Context, downloadDir string, firmwar
...
+		client,
+		sourceURL,
+	}
+}
+
+// Download will download the given firmware into the given downloadDir,
+// and return the full path to the downloaded file.
+// The file will be downloaded from the sourceURL provided to the SourceOverrideDownloader
+// instead of the firmware's UpstreamURL.
+func (d *SourceOverrideDownloader) Download(ctx context.Context, downloadDir string, firmware *serverservice.ComponentFirmwareVersion) (string, error) {
```
Comment (joelrebel) [inline internal/vendors/downloader.go:399]: Can we have a test for this method added

**Review by joelrebel — COMMENTED** [review]
Thanks for a work here, this looks fine to me. I've left a few comments.

File: internal/vendors/downloader.go (hunk 1 above)
Comment (coffeefreak101) [reply internal/vendors/downloader.go]: Done

File: internal/vendors/downloader.go (hunk 2 above)
Comment (coffeefreak101) [reply internal/vendors/downloader.go]: Done

File: internal/vendors/downloader.go (hunk 3 above)
Comment (coffeefreak101) [reply internal/vendors/downloader.go:400]: Sure thing! Done

**Review by joelrebel — APPROVED** [review]

**Pull request merged**